### ISSUE TITLE
Add a blacklist option

### ISSUE THREAD

**Issue by jtsagata** [TheAssassin/AppImageLauncher#338]
Add an option "Never integrate" besides "Run" and "Run and integrade".  

This will solve some issues with Apps like JetBrainsToolbox, or PureRef. The latest came also as a deb file, but the deb contains just an AppImage. So moving is impossible.  

An easy option to remove an item from the blacklist will be also nice to have.

**Comment by DASPRiD** [comment]
Same problem with some applications which are shipped as app images on steam. Right now I always have to click "Run once" each time I start them.

**Comment by TheAssassin** [comment]
@DASPRiD AppImages shipped on Steam should probably just declare that they don't want to be integrated. Please tell them to add `X-AppImage-Integrate=false` to their desktop file.

Just out of curiosity, who's shipping AppImages on Steam?

**Comment by DASPRiD** [comment]
@TheAssassin That'd be OpenVR-Advancd-Settings. I gonna make a pull requests against them :)

**Comment by username223** [comment]
@TheAssassin I'm a dev from [OpenVR-AdvancedSetting](https://github.com/OpenVR-Advanced-Settings/OpenVR-AdvancedSettings). We're shipping AppImages because the Steam Runtime are shipping hilariously out of date packages, that simply won't compile our application without massive changes anyway. Long term, the solution would probably to statically link as much as possible but that requires a massive amount of work and time, so we're doing AppImages for now.

We got the idea from Krita (https://github.com/ValveSoftware/steam-runtime/issues/189) who are doing the same thing. I don't know if Krita are still shipping AppImages on Steam, but it seems to work for us.

**Comment by Ahmed-E-86** [comment]
Also, `wine-launcher` doesn't launch, because of AppImageLauncher

**Comment by theblackhole** [comment]
> @DASPRiD AppImages shipped on Steam should probably just declare that they don't want to be integrated. Please tell them to add `X-AppImage-Integrate=false` to their desktop file.
> 
> Just out of curiosity, who's shipping AppImages on Steam?

@TheAssassin I have this issue with jetbrains toolbox, it always prompt me to integrate the AppImage but the app doesn't need it since it does it own thingand is already integrated. I added `X-AppImage-Integrate=false` to the .desktop file but it still opens AppImageLauncher everytime.
I want to give you my AppImageLauncher version but I don't know how to get it. The settings doesn't show it and neither `appimagelauncherd --version`, `appimagelauncherd -v`, `appimagelauncherd version`, `appimagelauncherd --help`, `appimagelauncherd -h` works...

**Comment by TheAssassin** [comment]
> @TheAssassin I have this issue with jetbrains toolbox, it always prompt me to integrate the AppImage but the app doesn't need it since it does it own thingand is already integrated. I added X-AppImage-Integrate=false to the .desktop file but it still opens AppImageLauncher everytime.

The author of the AppImage must add it to the desktop file inside the AppImage. Changes to the integrated one will be overwritten.

> I want to give you my AppImageLauncher version but I don't know how to get it. The settings doesn't show it and neither appimagelauncherd --version, appimagelauncherd -v, appimagelauncherd version, appimagelauncherd --help, appimagelauncherd -h works...

```sh
> AppImageLauncher --appimagelauncher-version

**Comment by theblackhole** [comment]
> The author of the AppImage must add it to the desktop file inside the AppImage. Changes to the integrated one will be overwritten.

Oh I see, thanks! Is there a way to do it myself? Because I use it everyday (it autostarts) so it's quite annoying (I don't want to disable AppImageLauncher completely but I might have to because of this)

> AppImageLauncher --appimagelauncher-version

I would have never thought of this, maybe consider adding a more standard `--version` option in the future ? (+ mention in the docs, to help you with issue reports and help us tracking the version ?)

**Comment by TheAssassin** [comment]
> I would have never thought of this, maybe consider adding a more standard --version option in the future ? (+ mention in the docs, to help you with issue reports and help us tracking the version ?)

Long overdue, please file a new issue.

>  Is there a way to do it myself?

First, [extract the AppImage's contents](https://docs.appimage.org/user-guide/run-appimages.html#extract-the-contents-of-an-appimage). You'll find the AppDir in `squashfs-root`. Then, patch the desktop file in the AppDir root, as well as in `squashfs-root/usr/share/applications/`.

Then, repack with [appimagetool](https://github.com/AppImage/AppImageKit/releases/latest):

```sh
> wget .../appimagetool-<arch>.AppImage
> chmod +x appimagetool*.AppImage
> ./appimagetool-<arch>.AppImage squashfs-root/
```

**Comment by theblackhole** [comment]
> First, extract the AppImage's contents. You'll find the AppDir in squashfs-root. Then, patch the desktop file in the AppDir root, > as well as in squashfs-root/usr/share/applications/.

> Then, repack with appimagetool:

> wget .../appimagetool-<arch>.AppImage
> chmod +x appimagetool*.AppImage
> ./appimagetool-<arch>.AppImage squashfs-root/

Thank you very much, I will try that

**Comment by martin-rueegg** [comment]
Have reported this to JetBrains: [Request #2992754](https://intellij-support.jetbrains.com/hc/en-us/requests/2992754).

But would still be great to be able to add exceptions to the [config file](https://github.com/TheAssassin/AppImageLauncher/wiki/Configuration). Even without UI option, but just documented.

**Comment by ebullient** [comment]
It would be nice if we could give AppImageLauncher a list of app images to ignore -- because going to each AppImage and re-packing (when jetbrains, e.g. will update itself) isn't really workable.

**Comment by smac89** [comment]
Yes a blacklist would be good way to deal with some of those stubborn appimages. https://github.com/TheAssassin/AppImageLauncher/issues/239#issuecomment-612572173

**Comment by darkporpoise** [comment]
Solution that worked for me for JetBrains toolbox on Login:
1) Edit .desktop file used for AutoStart (I use vi but nano etc works)
`vi ~/.config/autostart/jetbrains-toolbox.desktop`
2) Change `Exec` line to set `env`
From:
`Exec=/home/<user>/.local/share/JetBrains/Toolbox/bin/jetbrains-toolbox --minimize`
To:
`Exec=env APPIMAGELAUNCHER_DISABLE=TRUE /home/<user>/.local/share/JetBrains/Toolbox/bin/jetbrains-toolbox --minimize`

(Remember to change `<user>` to match your own)
3) Remove write privileges. This may not be necessary but mine seemed to revert back after a reboot otherwise
`chmod -w ~/.config/autostart/jetbrains-toolbox.desktop`

4) (optional) do the same thing for the applications .desktop file.
`vi ~/.local/share/applications/jetbrains-toolbox.desktop`

**Comment by TheAssassin** [comment]
These paths look quite constant... I am tempted to implement a specific exception...

**Comment by darkporpoise** [comment]
It's apparently fixed in the next release of JetBrains Toolbox (1.20). I think by fixed they mean `X-AppImage-Integrate=false` but it's a little unclear from the comments. I also think they are planning to make it so that the Toolbox can be run from arbitrary locations which would mean it could be integrated by AppImageLauncher.

https://youtrack.jetbrains.com/issue/TBX-3948

**Comment by Syzygianinfern0** [comment]
> Solution that worked for me for JetBrains toolbox on Login:
> 
> 1. Edit .desktop file used for AutoStart (I use vi but nano etc works)
>    `vi ~/.config/autostart/jetbrains-toolbox.desktop`
> 2. Change `Exec` line to set `env`
>    From:
>    `Exec=/home/<user>/.local/share/JetBrains/Toolbox/bin/jetbrains-toolbox --minimize`
>    To:
>    `Exec=env APPIMAGELAUNCHER_DISABLE=TRUE /home/<user>/.local/share/JetBrains/Toolbox/bin/jetbrains-toolbox --minimize`

Thank you! Used this solution for [Joplin](https://joplinapp.org/) and works splendidly. Needed it due to a weird way how Joplin manages updates and hence giving duplicate desktop entries.

**Comment by maverick74** [comment]
Any news on this?

**Comment by MMachado05** [comment]
It looks like the Joplin side of things moved on without adding an AppImageLauncher suppression option, so could this be picked back up again? It's still a bit of a hassle to always have to click "Run Once" whenver I want to take class notes...

**Comment by BrainStone** [comment]
It seems to recently have regressed into no longer setting `X-AppImage-Integrate=false`. I have [reported that](https://youtrack.jetbrains.com/issue/TBX-14579), but yeah having a blocklist would be nice.
